src/Workspaces/CSharp/Portable/CodeGeneration/CSharpCodeGenerationHelpers.cs

```diff
@@ -21,6 +20,67 @@ namespace Microsoft.CodeAnalysis.CSharp.CodeGeneration
 {
     internal static class CSharpCodeGenerationHelpers
     {
+        public static void AdjustNullableAnnotationByAttributes(ITypeSymbol type, ref ImmutableArray<AttributeData> attributes, ref NullableAnnotation nullableAnnotation, bool isParameter)
+        {
+            if (type.IsReferenceType)
+            {
+                if (!(type is ITypeParameterSymbol typeParameter) ||
+                    typeParameter.ReferenceTypeConstraintNullableAnnotation == NullableAnnotation.NotAnnotated)
+                {
+                    // For [AllowNull]/[MaybeNull], add the nullable annotation.
+                    // ([AllowNull] string?) -> (string?)
+                    // ([AllowNull] string)  -> (string?)
+                    var newAttributes = attributes.RemoveAll(IsAllowNullOrMaybeNullAttribute);
+                    if (newAttributes.Length != attributes.Length)
+                    {
+                        attributes = newAttributes;
+                        if (nullableAnnotation == NullableAnnotation.NotAnnotated)
+                        {
+                            nullableAnnotation = NullableAnnotation.Annotated;
+                        }
+                    }
+
+                    if (isParameter)
+                    {
+                        // The compiler does not allow parameter nullabilities to differ.
+                        // - ([DisallowNull] string?) -> (string)
+                        // + ([DisallowNull] string?) -> ([DisallowNull] string?)
+                        // ([DisallowNull] string)  -> (string)
```

Comment:
i don't really understand waht these comments mean.  what does the `-/+` indicator mean?

Reply:
```diff
-before/actual
+after/expected
```

The diff format, as you can see on GitHub.

Reply:
yeah, that's not ata ll clear to me.  i'm not seeing how it applies here either.  i would just say "when the user writes ... we want to produce ..."  and enumerate the before/after for each case.